handshake/handshake.go

```diff
@@ -0,0 +1,225 @@
+package handshake
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+
+	"github.com/renproject/aw/protocol"
+	"golang.org/x/crypto/sha3"
+)
+
+type Handshaker interface {
+	// Handshake with a remote server by initiating, and then interactively
+	// completing, a handshake protocol. The remote server is accessed by
+	// reading/writing to the `io.ReaderWriter`.
+	Handshake(ctx context.Context, rw io.ReadWriter) error
+
+	// AcceptHandshake from a remote client by waiting for the initiation of,
+	// and then interactively completing, a handshake protocol. The remote
+	// client is accessed by reading/writing to the `io.ReaderWriter`.
+	AcceptHandshake(ctx context.Context, rw io.ReadWriter) error
+}
+
+type handshaker struct {
+	signerVerifier protocol.SignerVerifier
+}
+
+func New(signerVerifier protocol.SignerVerifier) Handshaker {
+	return &handshaker{signerVerifier: signerVerifier}
+}
+
+func (hs *handshaker) Handshake(ctx context.Context, rw io.ReadWriter) error {
+	buf := new(bytes.Buffer)
+	rsaKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	// Write initiator's public key
+	if err := writePubKey(buf, rsaKey.PublicKey); err != nil {
+		return err
+	}
+	if err := hs.sign(buf, rw); err != nil {
+		return err
+	}
+
+	// Read responder's public key and challenge
+	if err := hs.verify(rw, buf); err != nil {
```

Comment:
This method could be named better. What exactly is being verified here, and how is the buffer being used? Same goes for most of the other private helper functions in this file. Their names are too generic. It is better to have longer, more descriptive, names especially for private helpers (that are not required to have accompanying comments).